dipy/reconst/ivim.py

```diff
@@ -0,0 +1,383 @@
+    else:
+        jacobian = None
+    for vox in range(flat_data.shape[0]):
+        res = Optimizer(sum_sq,
+                        flat_x0[vox],
+                        args=(bvals, flat_data[vox]), bounds=bounds,
+                        tol=tol, method=algorithm, jac=jacobian,
+                        options={'gtol': gtol, 'ftol': ftol, 'eps': eps, 'maxiter': maxiter})
+        ivim_params[vox, :4] = res.xopt
+        result += [res]
+    return result
+
+
+def _leastsq(flat_data, bvals, flat_x0, ivim_params):
+    """Use minimize for finding ivim_params"""
+    result = []
+    for vox in range(flat_data.shape[0]):
+        res = leastsq(_ivim_error,
+                      flat_x0[vox],
+                      args=(bvals, flat_data[vox]))
+        ivim_params[vox, :4] = res[0]
+        result += [res]
+    return result
+
+
+def two_stage(data, gtab, x0,
+              split_b, jac, bounds, tol,
+              routine, algorithm, gtol, ftol, eps, maxiter):
+    """
+    Fit the ivim params using a two stage fit
+
+    Parameters
+    ----------
+    data : array ([X, Y, Z, ...], g)
+        Data or response variables holding the data. Note that the last
+        dimension should contain the data. It makes no copies of data.
+
+    jac : bool
+        Use the Jacobian? Default: False
+
+    Returns
+    -------
+    ivim_params: the S0, f, D_star, D value for each voxel.
+
+    """
+    flat_data = data.reshape((-1, data.shape[-1]))
+    flat_x0 = x0.reshape(-1, x0.shape[-1])
+    # Flatten for the iteration over voxels:
+    bvals = gtab.bvals
+    ivim_params = np.empty((flat_data.shape[0], 4))
+
+    flat_x0[..., 0] = flat_data[..., 0]
+    D_guess = get_D_guess(flat_data, gtab, split_b)
+    S0_hat = get_S0_guess(flat_data, gtab, split_b)
+    f_guess = 1 - S0_hat / flat_data[..., 0]
+
+    flat_x0[..., 1] = f_guess
+    flat_x0[..., 3] = D_guess
+
+    if routine == 'minimize':
+        result = _minimize(flat_data, bvals, flat_x0, ivim_params,
+                           bounds, tol, jac, algorithm, gtol, ftol, eps, maxiter)
+    elif routine == "leastsq":
+        result = _leastsq(flat_data, bvals, flat_x0, ivim_params)
+    ivim_params.shape = data.shape[:-1] + (4,)
+    return ivim_params
+
+
+def get_D_guess(data, gtab, split_b):
+    bvals_ge_split = gtab.bvals[gtab.bvals > split_b]
+    bvecs_ge_split = gtab.bvecs[gtab.bvals > split_b]
+    gtab_ge_split = gradient_table(bvals_ge_split, bvecs_ge_split.T)
+    tensor_model = TensorModel(gtab_ge_split)
+    tenfit = tensor_model.fit(data[..., gtab.bvals > split_b])
+    D_guess = mean_diffusivity(tenfit.evals)
+    return D_guess
+
+
+def get_S0_guess(data, gtab, split_b):
+    bvals_ge_split = gtab.bvals[gtab.bvals > split_b]
+    bvecs_ge_split = gtab.bvecs[gtab.bvals > split_b]
+    gtab_ge_split = gradient_table(bvals_ge_split, bvecs_ge_split.T)
+    tensor_model = TensorModel(gtab_ge_split)
+    tenfit = tensor_model.fit(data[..., gtab.bvals > split_b])
```

Comment:
It seems that you are fitting the tensor model twice here. Once to guess the value of D and once to guess the value of S0. Should be faster to do only one model fit.